fixes #26918 - add a redis caching backend

timogoebel opened the pull request:
[no ci]

This will need packaging and changes in the puppet modules.

timogoebel committed:
fixes #26918 - add a redis caching backend

lzap commented on bundler.d/redis.rb:
```diff
@@ -0,0 +1,3 @@
+group :redis do
+  gem 'redis', '~> 4.0'
```
Do you have a particular reason why to stay on 4.0 instead of 4.1 which is the current stable? https://rubygems.org/gems/redis Perhaps even ~> 4 could be ok.

ekohl replied:
My understanding is that this means `>= 4.0 < 5.0` while `~> 4` is equivalent to `>= 4.0`. I think it's correct.

lzap commented on config/settings.yaml.example:
```diff
@@ -127,3 +127,18 @@
+#
+# Configure Connections to Redis
+# :redis_hosts:
+#   - redis://localhost:6379/0/cache
```
Can you remove "/0/cache" portion please? This looks confusing, the "/0" is probably fine according to specs (db number - http://www.iana.org/assignments/uri-schemes/prov/redis) however the "/cache" part does not make sense. I am going to test it anyway but it's like confusing.

timogoebel replied:
Yeah, you're right. The `/cache` specifies the namespace when using https://github.com/redis-store. But the rails implementation needs a separate key.

lzap reviewed (changes requested):
Looks good, few comments. I can do packaging for you, however I'd appreciate someone doing the installer part.

ehelms commented on config/settings.yaml.example:
```diff
@@ -127,3 +127,18 @@
 # :dynflow:
 #   :pool_size: 5
+
+#
```
What do you think about creating a nested set of cache options to namespace them? Something akin to:

```
:cache_store:
  :type: redis
  :namespace: foreman
  :compress: true
  :hosts:
    - redis://localhost:8479
```

If some of these options are quite redis specific, I think that could still live nested under something like `redis_options`.

lzap replied:
Actually, having Redis available opens up many new possibilities. I am thinking about exploring queuing for hooks or facts cache for faster updates of facts. But these instances should be probably separate, therefore this nicely fits this if I want to add new key `:facts_cache_store` in the future. Can you rename it to `:rails_cache_store` instead to differentiate it clearly?

timogoebel replied:
I had such a structure in the initial implementation and changed it because I thought we should probably just specify a single redis instance. But I agree, we should use a separate configuration.

ezr-ondrej replied:
I would add, that we are thinking about abusing redis for few more purposes than just caching and it would be nice to share the configuration of redis throughout all of its uses. Would it make sense?

timogoebel replied:
I‘m fine with both.

ehelms commented on config/application.rb:
```diff
@@ -153,7 +153,14 @@ class Application < Rails::Application
-    config.cache_store = :file_store, Rails.root.join("tmp", "cache")
+    if SETTINGS[:cache_store] == 'redis'
+      args = [:redis_cache_store]
+      Array.wrap(SETTINGS[:redis_hosts]).each { |h| args << h }
```
Is there a default host that is supported out of the box without having to configure it? If not could we add a default localhost if none other is provided?

timogoebel replied:
Rails defaults to `redis://localhost:6379/0`.

lzap replied:
Works out of box, tested.

ekohl reviewed (commented):
If we continue the pattern, this means we'll create a `foreman-redis` subpackage that has the correct dependency. The installer would then have options for that. We can have a look at including redis itself as well.

ohadlevy commented:
simple way to get redis running via docker-compose at https://github.com/ohadlevy/foreman/commit/67df19e6ed6bc2acb218110016a2a7f21226df5d

ohadlevy commented on config/application.rb:
```diff
@@ -153,7 +153,14 @@ class Application < Rails::Application
+    if SETTINGS[:cache_store] == 'redis'
+      args = [:redis_cache_store]
+      Array.wrap(SETTINGS[:redis_hosts]).each { |h| args << h }
+      args << { namespace: 'foreman' }.merge(SETTINGS[:redis_cache_options] || {})
```
could we add support to define it via ENV variable as well? something like
```rb
config.cache_store = :redis_cache_store, { url: ENV['REDIS_URL'] }
```

ohadlevy replied:
another option, would be to allow defining the entire yaml as a env variable, something like
```diff
diff --git a/config/settings.rb b/config/settings.rb
index 7640ed7e5..45141b3da 100644
--- a/config/settings.rb
+++ b/config/settings.rb
@@ -5,6 +5,7 @@ root = File.expand_path(File.dirname(__FILE__) + "/..")
 settings_file = Rails.env.test? ? 'config/settings.yaml.test' : 'config/settings.yaml'
 
 SETTINGS.merge! YAML.load(ERB.new(File.read("#{root}/#{settings_file}")).result) if File.exist?(settings_file)
+SETTINGS.merge! YAML.load(ENV['FOREMAN_SETTINGS']) if ENV['FOREMAN_SETTINGS']
 SETTINGS[:version] = Foreman::Version.new
 
```
WDYT?

ekohl replied:
I don't think a YAML settings hash is a great idea since it's very hard to manipulate.

How about shipping a `config/settings.yaml` file inside the docker environment (content completely made up):
```erb
---
:redis_cache_store: <%= ENV['REDIS_CACHE_STORE'] || 'my_store' %>
```

timogoebel replied:
That would be the easiest solution, but I think we should be able to parse the ENV variables as well and store them in a settings hash. Ideally, we would also validate the boot settings file to make sure only allowed variables are set.

timogoebel replied:
I've filed a PR at https://github.com/theforeman/foreman/pull/6826 that adds proper support for defining settings in env variables.

ohadlevy reviewed (changes requested):
thanks!

could you please add support for sessions too? e.g.
```rb
config.session_store :cache_store, key: ENV['APP_SESSION_KEY']
```
(taken from https://medium.com/@kirill_shevch/configuration-cache-and-rails-session-store-with-redis-b3ce6f64d1fc)

this should allow us to deprecate memcache plugin as welldepericate

lzap commented:
I suggest to design it it a way that session and cache can be defined as two separate databases. This way you can define how much RAM you willing to allocate for each other.

lzap commented:
Also I suggest using UNIX socket by default as it appears to give much more performance at almost zero cost: https://redis.io/topics/benchmarks

(Obviously for the standard deployment, not for containers where TCP is the only option I think.)

lzap commented:
Packaging PR: https://github.com/theforeman/foreman-packaging/pull/3831

timogoebel commented:
> could you please add support for sessions too?

Can we please add that in a separate PR? I believe we currently use the (encrypted) cookie store which I prefer over redis.

timogoebel commented:
> Also I suggest using UNIX socket by default

I think we should do that but set this in the puppet code/installer.

timogoebel commented:
Updated according to your comments, thanks for all the feedback: :heart:

ohadlevy commented on config/application.rb:
```diff
@@ -153,7 +153,20 @@ class Application < Rails::Application
+    rails_cache_settings = SETTINGS[:rails_cache_store]
+    if rails_cache_settings && rails_cache_settings[:type] == 'redis'
+      args = [:redis_cache_store]
+      if ENV['RAILS_CACHE_STORE_HOSTS'].present?
```
can you swap the condition that if redis env is defined the cache is also set? my aim is to expose it via docker env without the settings file..

timogoebel replied:
@ohadlevy: Sure, like this?

ekohl replied:
I have my concerns about adding environment variables for configuration. Not against it in general but I think we can end up with a very mixed and undocumented set of env vars.

Can't we read env vars in the settings layer and then just use the `SETTINGS` class in here? It at least keeps a clean layer in between.

ohadlevy replied:
yes, just wondering if the name of the variable makes sense, its redis specific vs rails cache store in general?also, do we really assume we have multiple cache servers? (I can see multiple redis with persisitent data, less for cache) so all in all in my opinion, I think ```REDIS_URL``` is easier and perhaps can add multi host with ```REDIS_URL='redis://one redis://two'```...

ohadlevy replied:
> I have my concerns about adding environment variables for configuration. Not against it in general but I think we can end up with a very mixed and undocumented set of env vars.
> 
> Can't we read env vars in the settings layer and then just use the `SETTINGS` class in here? It at least keeps a clean layer in between.

like https://github.com/theforeman/foreman/pull/6811#discussion_r289146370?

timogoebel replied:
I think we should first decide if we want to configure a default redis connection that will be used for every use case or allow a user to configure separate instances. This will answer if we need to prefix the variable with `RAILS_CACHE_STORE_` or not.
I think we should definitely allow setting multiple hosts.

> Can't we read env vars in the settings layer and then just use the SETTINGS class in here?
@ekohl: That would be great, but we'd need to cast these variables. SETTINGS basically just loads a YAML-file, see https://github.com/theforeman/foreman/blob/develop/config/boot_settings.rb. We can, of course, hardcode all supported settings there. Maybe that's not such a bad idea.

ekohl replied:
I'd be in favor of merging this without environment variables (since Redis is useful on its own) and then opening the bikeshedding discussion.

timogoebel commented:
I've removed the env variable for now and renamed the setting from `hosts` to `urls`.

```diff
diff --git a/config/application.rb b/config/application.rb
index 65e6df348..010aafa19 100644
--- a/config/application.rb
+++ b/config/application.rb
@@ -154,14 +154,10 @@ module Foreman
     # enables in memory cache store with ttl
     # config.cache_store = TimedCachedStore.new
     rails_cache_settings = SETTINGS[:rails_cache_store]
-    if (rails_cache_settings && rails_cache_settings[:type] == 'redis') || ENV['RAILS_CACHE_STORE_HOSTS'].present?
+    if (rails_cache_settings && rails_cache_settings[:type] == 'redis')
       args = [:redis_cache_store]
-      if ENV['RAILS_CACHE_STORE_HOSTS'].present?
-        redis_hosts = ENV['RAILS_CACHE_STORE_HOSTS'].split(',').compact.map(&:strip)
-      else
-        redis_hosts = Array.wrap(rails_cache_settings[:hosts])
-      end
-      redis_hosts.each { |h| args << h }
+      redis_urls = Array.wrap(rails_cache_settings[:urls])
+      redis_urls.each { |url| args << url }
       args << { namespace: 'foreman' }.merge(rails_cache_settings[:options] || {})
       config.cache_store = args
     else
diff --git a/config/settings.yaml.example b/config/settings.yaml.example
index fae462775..bee9a74fe 100644
--- a/config/settings.yaml.example
+++ b/config/settings.yaml.example
@@ -134,7 +134,7 @@
 #
 # :rails_cache_store:
 #   :type: redis
-#   :hosts:
+#   :urls:
 #    - redis://localhost:8479/0
 #   :options:
 #     :compress: true
```

timogoebel commented:
I've filed https://projects.theforeman.org/issues/26955 for allowing to set the boot variables via environment variables.

timogoebel commented:
@dgoetz: Does any of your trainees want to learn more puppet and add installer support for this? 👼

dgoetz commented:
@timogoebel Yes, if it is not urgent, because earliest slot would be calendar week 27 or 29. 

It would be adding a option to enable redis, the settings in settings.yaml, introducing [redis](https://github.com/voxpupuli/puppet-redis) as new dependeny and using it to set redis up matching the settings, defaulting to unix socket. Anything missing?

ohadlevy commented:
@lzap @ekohl this looks fine for me, would you remove your nack? thanks.

ares reviewed (commented):
Awesome! It would be absolutely great to write a short blog post about the setup or update the manual. Also probably listing this in 1.23 release notes and 1.23 roadmap on discourse. Last but not least, we should update foreman_memcache readme there's this alternative. Is there someone willing to take these actions?

timogoebel commented:
> It would be absolutely great to write a short blog post about the setup or update the manual.

I would wait until we have full support in the installer.

> Last but not least, we should update foreman_memcache readme there's this alternative.

That's something I can do.

ekohl commented on config/application.rb:
```diff
@@ -153,7 +153,16 @@ class Application < Rails::Application
 
     # enables in memory cache store with ttl
```
Nit: IMHO we should either remove this comment or support it. Probably the former :)

ekohl commented on config/application.rb:
```diff
@@ -153,7 +153,16 @@ class Application < Rails::Application
     # config.cache_store = TimedCachedStore.new
-    config.cache_store = :file_store, Rails.root.join("tmp", "cache")
+    rails_cache_settings = SETTINGS[:rails_cache_store]
+    if (rails_cache_settings && rails_cache_settings[:type] == 'redis')
```
How about explicitly checking if it's a hash so a user doesn't accidentally use a string?

```suggestion
    if (rails_cache_settings.is_a?(Hash) && rails_cache_settings[:type] == 'redis')
```

I also wouldn't mind hard failing if it's something else so we don't silently fall back to files.

lzap replied:
I don't think it's necessary, the current code will make the app to fail if user accidentally set it to string which is actually better than silently ignoring it.

ekohl reviewed (commented):
Is there an easy way for the user to verify Redis is actually used as a cache? Perhaps an info log message during startup?

timogoebel commented:
We can‘t log anything as logging is not initialized at that time. But we could maybe do that later.

ekohl commented:
My biggest concern is that the user things they've configured caching, but it's not actually used. Since it currently silently falls back and there's no way that they can see it, it's a serious risk for a lot of frustration (as we've learned from configuring the smart proxy). I think a startup log message about the caching backend is good enough to mitigate that.

ohadlevy commented:
> My biggest concern is that the user things they've configured caching, but it's not actually used. Since it currently silently falls back and there's no way that they can see it, it's a serious risk for a lot of frustration (as we've learned from configuring the smart proxy). I think a startup log message about the caching backend is good enough to mitigate that.

my suggestion (not in this PR) is to:
1. Log on cache failures (with the right error level)
2. Send a notification  (using the notification drawer) to administrators that something is not right
3. replace the about page with a status page and display something regarding it there.

lzap commented:
I suggest to document in docs how to configure cache and also how to test it really works. Additionally, we can write a foreman-maintain health check that will easily verify that the cache is operating correctly. I don't think that a warning message in logs helps at all in this situation.

lzap reviewed (dismissed):
I think we all expressed various ideas for the future, but the patch is good enough and simple: "add redis for rails cache". Anything else we can refactor. I am good, going to test this and merge once packages are ready.

lzap commented:
It does not work, you send three arguments:

    config.cache_store = :redis_cache_store, redis://localhost:8479/0, {:namespace=>"foreman", :compress=>true}

But in documentation I read it should be just two: https://guides.rubyonrails.org/caching_with_rails.html#activesupport-cache-rediscachestore

The error I experience is: /home/lzap/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/activesupport-5.2.1/lib/active_support/cache/redis_cache_store.rb:174:in `initialize': wrong number of arguments (given 1, expected 0) (ArgumentError)

timogoebel commented:
@lzap: Thanks, that's a little embarrassing. I still had my initial implementation of this (as a plugin) enabled, so no matter how much wrong code I actually configured in core: It always worked.

Fixed the issue and added some simple log statements to make @ekohl happy (cause he is - as always 
- right).

lzap reviewed (approved):
No problem it happens to all of us! Lemme give this a quick test. Katello failures unrelated:

    Actions::Pulp3::CopyAllUnitsTest.test_create
    Actions::Pulp3::GenerateMetadataTest.test_generate_with_source_repo
